File: go/apps/dialogue/tests/test_vumi_app.js
var assert = require('assert');

var _ = require('lodash');
var vumigo = require('vumigo_v02');
var fail = vumigo.test_utils.fail;
var AppTester = vumigo.AppTester;

var app = require('../vumi_app');
var DialogueApp = app.DialogueApp;
var dummy_polls = require('./dummy_polls');

describe("app", function() {
    describe("DialogueApp", function() {
        var poll;
        var tester;

        beforeEach(function() {
            poll = _.cloneDeep(dummy_polls.simple_poll);

            tester = new AppTester(new DialogueApp())
                .setup.config.app({name: 'dialogue_app'})
                .setup.config({poll: poll}, {json: false})
                .setup.user.addr('+27123');
        });

        function extend_poll(poll_extend) {
            return tester.setup.config({
                poll: _.extend(poll, poll_extend)
            },
            {json: false});
        }

        it("should throw an error if an unknown state type is encountered",
        function() {
            return extend_poll({states: [{type: 'unknown'}]})
                .run()
                .catch(function(e) {
                    assert(e instanceof Error);
                    assert.equal(
                        e.message,
                        "Unknown dialogue state type: 'unknown'");
                });
        });

        it("should stay on the same state if a state has no next state",
        function() {
            return extend_poll({connections: []})
                .setup.user.state('choice-1')
                .input('1')
                .check.user.state('choice-1')
                .run();
        });

        describe("when the user enters a choice state", function() {
            it("should display the state's question", function() {
                return tester
                    .start()
                    .check.user.state('choice-1')
                    .check.reply([
                        "What is your favourite colour?",
                        "1. Red",
                        "2. Blue",
                        "3. Green",
                        "4. Black",
                        "5. Yellow"
                    ].join('\n'))
                    .run();
            });

            it("should move the user to the next state on valid input",
            function() {
                return tester
                    .setup.user.state('choice-1')
                    .input('1')
                    .check.user.state('freetext-1')
                    .run();
            });

            it("should stay on the same state on invalid input", function() {
                return tester
                    .setup.user.state('choice-1')
                    .input('23')
                    .check.user.state('choice-1')
                    .run();
            });

            it("should store the user's answer", function() {
                return tester
                    .inputs(
                        null, '1', 'foo',
                        null, '2', 'bar',
                        null, '1', 'baz',
                        null, '2', 'quux')
                    .check(function(api) {
                        var contact = _.find(api.contacts.store, {
                            msisdn: '+27123'
                        });

                        assert.equal(contact.extra['message-1'], 'value-2');
                        assert.equal(contact.extra['message-1-1'], 'value-1');
                        assert.equal(contact.extra['message-1-2'], 'value-2');
                        assert.equal(contact.extra['message-1-3'], 'value-1');
                        assert.equal(contact.extra['message-1-4'], 'value-2');
                    })
                    .run();
            });

            it("should only accept number based answers if asked", function() {
                return extend_poll({accept_labels: false})
                    .setup.user.state('choice-1')
                    .input('Red')
                    .check.user.state('choice-1')
                    .run();
            });

            it("should accept label and number based answers if asked",
            function() {
                return extend_poll({accept_labels: true})
                    .setup.user.state('choice-1')
                    .input('Red')
                    .check.user.state('freetext-1')
                    .run();
            });
        });

        describe("when the user enters a freetext state", function() {
            it("should display the state's question", function() {
                return tester
                    .setup.user.state('choice-1')
                    .input('1')
                    .check.user.state('freetext-1')
                    .check.reply('What is your name?')
                    .run();
            });

            it("should move the user to the next state on valid input",
            function() {
                return tester
                    .setup.user.state('freetext-1')
                    .input('foo')
                    .check.user.state('end-1')
                    .run();
            });

            it("should store the user's answer", function() {
                return tester
                    .inputs(
                        null, '1', 'foo',
                        null, '1', 'bar',
                        null, '1', 'baz',
                        null, '1', 'quux')
                    .check(function(api) {
                        var contact = _.find(api.contacts.store, {
                            msisdn: '+27123'
                        });

                        assert.equal(contact.extra['message-3'], 'quux');
                        assert.equal(contact.extra['message-3-1'], 'foo');
                        assert.equal(contact.extra['message-3-2'], 'bar');
                        assert.equal(contact.extra['message-3-3'], 'baz');
                        assert.equal(contact.extra['message-3-4'], 'quux');
                    })
                    .run();
            });
        });

        describe("when the user enters an end state", function() {
            it("should display the state's text", function() {
                return tester
                    .setup.user.state('choice-1')
                    .input('2')
                    .check.user.state('end-1')
                    .check.reply('Thank you for taking our survey')
                    .run();
            });

            it("should move to the start state next session if asked",
            function() {
                return extend_poll({poll_metadata: {repeatable: true}})
                    .setup.user.state('end-1')
                    .start()
                    .check.user.state('choice-1')
                    .run();
            });

            it("should not move to the start state next session if asked",
            function() {
                return extend_poll({poll_metadata: {repeatable: false}})
                    .setup.user.state('end-1')
                    .start()
                    .check.user.state('end-1')
                    .run();
            });
        });

        describe("when the user enters a group state", function() {
            it("should add the contact to the given group", function() {
                return tester
                    .setup(function(api) {
                        api.contacts.add({
                            msisdn: '+27123',
                            groups: ['group-2']
                        });
                    })
                    .setup.user.addr('+27123')
                    .setup.user.state('choice-1')
                    .input('1')
                    .check(function(api) {
                        var contact = api.contacts.store[0];

                        assert.deepEqual(contact.groups.sort(), [
                            'group-1',
                            'group-2'
                        ]);
                    })
                    .run();
            });

            it("should not change the contact's groups if none is assigned",
            function() {
                return tester
                    .setup(function(api) {
                      var state = _.find(api.config.store.poll.states, {
                        uuid: 'group-1'
                      });

                      state.group = null;
                    })
                    .setup(function(api) {
                        api.contacts.add({
                            msisdn: '+27123',
                            groups: ['group-2']
                        });
                    })
                    .setup.user.addr('+27123')
                    .setup.user.state('choice-1')
                    .input('1')
                    .check(function(api) {
                        var contact = api.contacts.store[0];
                        assert.deepEqual(contact.groups.sort(), ['group-2']);
                    })
                    .run();
            });

            it("should redirect to the state linked to the exit endpoint",
            function() {
                return tester
                    .setup.user.state('choice-1')
                    .input('1')
                    .check.user.state('freetext-1')
                    .check.reply('What is your name?')
                    .run();
            });
        });

        describe("when the user enters a send state", function() {
            it("should send a message over the given channel", function() {
                return tester
                    .setup.config.app({
                        endpoints: {Twitter: {delivery_class: 'twitter'}}
                    })
                    .setup(function(api) {
                        api.contacts.add({
                            msisdn: '+27123',
                            twitter_handle: '@me'
                        });
                    })
                    .setup.user.addr('+27123')
                    .setup.user.state('choice-1')
                    .input('3')
                    .check(function(api) {
                        assert.equal(api.outbound.store.length, 2);

                        var msg = api.outbound.store[0];
                        assert.equal(msg.to_addr, '@me');
                        assert.equal(msg.content, "Hello over other endpoint");
                    })
                    .run();
            });

            it("should redirect to the state linked to the exit endpoint",
            function() {
                return tester
                    .setup.config.app({
                        endpoints: {Twitter: {delivery_class: 'twitter'}}
                    })
                    .setup(function(api) {
                        api.contacts.add({
                            msisdn: '+27123',
                            twitter_handle: '@me'
                        });
                    })
                    .setup.user.state('choice-1')
                    .input('3')
                    .check.user.state('freetext-1')
                    .check.reply('What is your name?')
                    .run();
            });

            it("should throw an error if no channel type is found", function() {
                var state = _.find(poll.states, {uuid: 'send-1'});
                state.channel_type = '???';

                return tester
                    .setup(function(api) {
                        api.contacts.add({
                            msisdn: '+27123',
                            twitter_handle: '@me'
                        });
                    })
                    .setup.user.state('choice-1')
                    .input('3')
                    .run()
                    .then(fail, function(e) {
                        assert(e instanceof Error);

                        assert.equal(
                            e.message,
                            "No endpoint found for channel type '???'");
                    });
            });
        });

        describe("when the user enters a send state which is a httpjson type", function() {
            it("should successfully fetch the data from the url if the method is GET", function() {
                return tester
                    .setup(function(api) {
                        api.contacts.add({
                           msisdn: '+27123'
                        });
                        api.http.fixtures.add({
                            request: {
                                method: "GET", 
                                url: "www.foo.bar"
                            }
                        });
                    })
                    .setup.user.addr('+27123')
                    .setup.user.state('choice-1')
                    .input('4')
                    .check(function(api) {
                        assert.equal(api.http.fixtures.fixtures[0].responses[0].code, 200);
                    })
                    .run();
            });
            
            it("should successfully send the data to the url if the method is POST", function() {
                return tester
                    .setup.user.addr('+27123')                    
                    .setup.user.state('choice-1')
                    .setup(function(api) {
                        api.contacts.add({
                            msisdn: '+27123'
                        });

                        api.contacts.store[0].extra = { 
                            'message-1': 'value-5', 
                            'message-1-1': 'value-5' 
                        };

                        api.http.fixtures.add({
                            request: {
                                method: "POST", 
                                url: "www.foo.bar", 
                                data: JSON.stringify({
                                    user: {
                                        answers: { 
                                            'choice-1': 'value-5'
                                        }
                                    }, 
                                    contact: api.contacts.store[0], 
                                    conversation_key: poll.conversation_key
                                })
                            }
                        });

                        api.contacts.store[0].extra = {};
                    })
                    .input('5')
                    .check(function(api) {
                        assert.equal(api.http.fixtures.fixtures[0].responses[0].code, 200);
                    })
                    .run();
            });
        });
    });
});

File: go/apps/dialogue/tests/test_vumi_app.py
# -*- coding: utf-8 -*-

"""Tests for go.apps.dialogue.vumi_app"""

import os
import json
import pkg_resources

from twisted.internet.defer import inlineCallbacks, returnValue

from vxsandbox.utils import find_nodejs_or_skip_test
from vxsandbox.tests.utils import DummyAppWorker
from vxsandbox.resources.tests.utils import ResourceTestCaseBase

from vumi.tests.helpers import VumiTestCase
from vumi.tests.utils import LogCatcher

from go.apps.dialogue.vumi_app import DialogueApplication, PollConfigResource
from go.apps.dialogue.utils import dialogue_js_config
from go.apps.dialogue.tests.dummy_polls import simple_poll
from go.apps.tests.helpers import AppWorkerHelper


class DummyDialogueAppWorker(DummyAppWorker):
    def __init__(self):
        super(DummyDialogueAppWorker, self).__init__()
        self.conv = None

    def conversation_for_api(self, api):
        return self.conv


class TestDialogueApplication(VumiTestCase):

    @inlineCallbacks
    def setUp(self):
        nodejs_executable = find_nodejs_or_skip_test(DialogueApplication)

        self.app_helper = self.add_helper(AppWorkerHelper(DialogueApplication))

        sandboxer_js = pkg_resources.resource_filename('vumi.application',
                                                       'sandboxer.js')
        node_path = os.environ['SANDBOX_NODE_PATH']  # Required to run tests.
        redis = yield self.app_helper.vumi_helper.get_redis_manager()
        self.kv_redis = redis.sub_manager('kv')
        self.app = yield self.app_helper.get_app_worker({
            'executable': nodejs_executable,
            'args': [sandboxer_js],
            'timeout': 10,
            'app_context': (
                "{require: function(m) {"
                " if (['moment', 'url', 'querystring', 'crypto', 'lodash',"
                " 'q', 'jed', 'libxmljs', 'zlib', 'vumigo_v01', 'vumigo_v02'"
                "].indexOf(m) >= 0) return require(m); return null;"
                " }, Buffer: Buffer}"
            ),
            'env': {
                'NODE_PATH': node_path,
            },
            'sandbox': {
                'config': {
                    'cls': 'go.apps.dialogue.vumi_app.PollConfigResource',
                },
                'contacts': {
                    'cls': 'go.apps.jsbox.contacts.ContactsResource',
                },
                'kv': {
                    'cls': 'vxsandbox.RedisResource',
                    'redis_manager': {'FAKE_REDIS': self.kv_redis},
                },
                'outbound': {
                    'cls': 'go.apps.jsbox.outbound.GoOutboundResource',
                },
            },
            'rlimits': {
                'RLIMIT_STACK': [2 * 1024 * 1024] * 2,
                'RLIMIT_AS': [256 * 1024 * 1024] * 2,
            },
        })

    @inlineCallbacks
    def setup_conversation(self, poll=None):
        group = yield self.app_helper.create_group(u'group')

        yield self.app_helper.create_contact(
            msisdn=u'+278312345670',
            twitter_handle=u'@0',
            groups=[group],
            name=u'Contact',
            surname=u'0')

        yield self.app_helper.create_contact(
            msisdn=u'+278312345671',
            twitter_handle=u'@1',
            groups=[group],
            name=u'Contact',
            surname=u'1')

        config = {"poll": poll or simple_poll()}
        conv = yield self.app_helper.create_conversation(
            config=config, groups=[group])
        returnValue(conv)

    def send_send_jsbox_command(self, conversation):
        return self.app_helper.dispatch_command(
            "send_jsbox",
            user_account_key=conversation.user_account.key,
            conversation_key=conversation.key,
            batch_id=conversation.batch.key,
        )

    @inlineCallbacks
    def test_send_dialogue_command(self):
        conv = yield self.setup_conversation()
        yield self.app_helper.start_conversation(conv)
        with LogCatcher(message='Switched to state:') as lc:
            yield self.send_send_jsbox_command(conv)
            self.assertEqual(lc.messages(),
                             ['Switched to state: choice-1'] * 2)
        msgs = self.app_helper.get_dispatched_outbound()
        for msg in msgs:
            self.assertEqual(msg["content"],
                             "What is your favourite colour?\n1. Red\n2. Blue")
            self.assertEqual(msg["in_reply_to"], None)
            go_metadata = msg["helper_metadata"]["go"]
            self.assertEqual(go_metadata["conversation_type"], "dialogue")
            self.assertEqual(go_metadata["conversation_key"], conv.key)
        self.assertEqual(sorted([m["to_addr"] for m in msgs]),
                         ["+278312345670", "+278312345671"])

    @inlineCallbacks
    def test_send_dialogue_command_delivery_class(self):
        poll = simple_poll()
        poll['poll_metadata']['delivery_class'] = 'twitter'
        conv = yield self.setup_conversation(poll=poll)
        yield self.app_helper.start_conversation(conv)
        with LogCatcher(message='Switched to state:') as lc:
            yield self.send_send_jsbox_command(conv)
            self.assertEqual(lc.messages(),
                             ['Switched to state: choice-1'] * 2)
        msgs = self.app_helper.get_dispatched_outbound()
        for msg in msgs:
            self.assertEqual(msg["content"],
                             "What is your favourite colour?\n1. Red\n2. Blue")
            self.assertEqual(msg["in_reply_to"], None)
            go_metadata = msg["helper_metadata"]["go"]
            self.assertEqual(go_metadata["conversation_type"], "dialogue")
            self.assertEqual(go_metadata["conversation_key"], conv.key)
        self.assertEqual(sorted([m["to_addr"] for m in msgs]),
                         ["@0", "@1"])

    @inlineCallbacks
    def test_user_message(self):
        conversation = yield self.setup_conversation()
        yield self.app_helper.start_conversation(conversation)
        with LogCatcher(message='Switched to state:') as lc:
            yield self.app_helper.make_dispatch_inbound(
                "hello", conv=conversation)
            self.assertEqual(lc.messages(),
                             ['Switched to state: choice-1'])
        [reply] = self.app_helper.get_dispatched_outbound()
        self.assertEqual(reply["content"],
                         "What is your favourite colour?\n1. Red\n2. Blue")

    @inlineCallbacks
    def test_ack(self):
        conversation = yield self.setup_conversation()
        yield self.app_helper.start_conversation(conversation)
        msg = yield self.app_helper.make_stored_outbound(
            conversation, "foo")
        with LogCatcher(message="Ignoring") as lc:
            yield self.app_helper.make_dispatch_ack(msg, conv=conversation)
        self.assertEqual(
            lc.messages(),
            ["Ignoring event for conversation: %s" % (conversation.key,)])

    @inlineCallbacks
    def test_delivery_report(self):
        conversation = yield self.setup_conversation()
        yield self.app_helper.start_conversation(conversation)
        msg = yield self.app_helper.make_stored_outbound(
            conversation, "foo")
        with LogCatcher(message="Ignoring") as lc:
            yield self.app_helper.make_dispatch_delivery_report(
                msg, conv=conversation)
        self.assertEqual(
            lc.messages(),
            ["Ignoring event for conversation: %s" % (conversation.key,)])

    @inlineCallbacks
    def test_send_message_command(self):
        conversation = yield self.setup_conversation()
        yield self.app_helper.start_conversation(conversation)
        msg_options = {
            'transport_name': 'sphex_transport',
            'from_addr': '666666',
            'transport_type': 'sphex',
            'helper_metadata': {'foo': {'bar': 'baz'}},
        }
        yield self.app_helper.dispatch_command(
            "send_message",
            user_account_key=conversation.user_account.key,
            conversation_key=conversation.key,
            command_data={
                "batch_id": conversation.batch.key,
                "to_addr": "123456",
                "content": "hello world",
                "msg_options": msg_options,
            })

        [msg] = self.app_helper.get_dispatched_outbound()
        self.assertEqual(msg.payload['to_addr'], "123456")
        self.assertEqual(msg.payload['from_addr'], "666666")
        self.assertEqual(msg.payload['content'], "hello world")
        self.assertEqual(msg.payload['transport_name'], "sphex_transport")
        self.assertEqual(msg.payload['transport_type'], "sphex")
        self.assertEqual(msg.payload['message_type'], "user_message")
        self.assertEqual(msg.payload['helper_metadata']['go'], {
            'user_account': conversation.user_account.key,
            'conversation_type': 'dialogue',
            'conversation_key': conversation.key,
        })
        self.assertEqual(msg.payload['helper_metadata']['foo'],
                         {'bar': 'baz'})

    @inlineCallbacks
    def test_process_command_send_message_in_reply_to(self):
        conversation = yield self.setup_conversation()
        yield self.app_helper.start_conversation(conversation)
        msg = yield self.app_helper.make_stored_inbound(conversation, "foo")
        yield self.app_helper.dispatch_command(
            "send_message",
            user_account_key=conversation.user_account.key,
            conversation_key=conversation.key,
            command_data={
                "batch_id": conversation.batch.key,
                "to_addr": "to_addr",
                "content": "foo",
                u'msg_options': {
                    u'transport_name': u'smpp_transport',
                    u'in_reply_to': msg['message_id'],
                    u'transport_type': u'sms',
                    u'from_addr': u'default10080',
                },
            })
        [sent_msg] = self.app_helper.get_dispatched_outbound()
        self.assertEqual(sent_msg['to_addr'], msg['from_addr'])
        self.assertEqual(sent_msg['content'], 'foo')
        self.assertEqual(sent_msg['in_reply_to'], msg['message_id'])


class TestPollConfigResource(ResourceTestCaseBase):
    # TODO: Make this resource stuff into a helper in vumi.
    app_worker_cls = DummyDialogueAppWorker
    resource_cls = PollConfigResource

    @inlineCallbacks
    def setUp(self):
        super(TestPollConfigResource, self).setUp()
        self.app_helper = self.add_helper(AppWorkerHelper(DialogueApplication))
        self.app = yield self.app_helper.get_app_worker({})
        yield self.create_resource({})

    @inlineCallbacks
    def test_get(self):
        conv = yield self.app_helper.create_conversation(
            config={'poll': simple_poll()})
        self.app_worker.conv = conv

        reply = yield self.dispatch_command('get', key='config')
        config = json.loads(reply['value'])

        self.assertEqual(config, dialogue_js_config(conv))
